banyand/property/repair.go

```diff
@@ -0,0 +1,947 @@
+	writer     *bufio.Writer
+	Slots      []*repairTreeBuilderSlot
+	currentOff int64
+}
+
+func newRepairTreeBuilder(groupFilePath string) (*repairTreeBuilder, error) {
+	if err := os.MkdirAll(filepath.Dir(groupFilePath), storage.DirPerm); err != nil {
+		return nil, fmt.Errorf("creating directory for repair tree file %s failure: %w", groupFilePath, err)
+	}
+	groupFile, err := os.OpenFile(groupFilePath, os.O_WRONLY|os.O_CREATE, storage.FilePerm)
+	if err != nil {
+		return nil, fmt.Errorf("opening repair tree file %s failure: %w", groupFilePath, err)
+	}
+	return &repairTreeBuilder{
+		Slots:     make([]*repairTreeBuilderSlot, 0),
+		rootHash:  sha512.New(),
+		groupFile: groupFile,
+		writer:    bufio.NewWriter(groupFile),
+	}, nil
+}
+
+func (r *repairTreeBuilder) appendSlot(slot *repairSlotFile) (err error) {
+	slotSha := slot.sha.Sum(nil)
+	shaValBytes := []byte(fmt.Sprintf("%x", slotSha))
+	_, err = r.rootHash.Write(shaValBytes)
+	if err != nil {
+		return fmt.Errorf("hashing root sha value failure: %w", err)
+	}
+	bytesSize, leafCount, err := slot.writeToAndDeleteSlotFile(r.writer)
+	if err != nil {
+		return fmt.Errorf("writing slot file %s to repair tree file failure: %w", slot.path, err)
+	}
+	r.Slots = append(r.Slots, &repairTreeBuilderSlot{
+		shaVal:         shaValBytes,
+		repairSlotFile: slot,
+		startOff:       r.currentOff,
+		leafCount:      leafCount,
+		writeLen:       bytesSize,
+	})
+	r.currentOff += bytesSize
+	return nil
+}
+
+func (r *repairTreeBuilder) build() (err error) {
+	leafNodesFinishedOffset := r.currentOff
+	slotNodesFinishedOffset := r.currentOff
+	var slotNodesLen int64
+	// write the slot nodes
+	var writedLen int
+	for _, slot := range r.Slots {
+		data := make([]byte, 0)
+		data = encoding.VarInt64ToBytes(data, int64(slot.slot))
+		data = encoding.EncodeBytes(data, slot.shaVal)
+		data = encoding.VarInt64ToBytes(data, slot.startOff)
+		data = encoding.VarInt64ToBytes(data, slot.leafCount)
+		data = append(data, repairFileNewLine)
+		writedLen, err = r.writer.Write(data)
+		if err != nil {
+			return fmt.Errorf("writing slot node to repair tree file failure: %w", err)
+		}
+		slotNodesLen += int64(writedLen)
+		slotNodesFinishedOffset += int64(writedLen)
+	}
+
+	// write the root node
+	data := make([]byte, 0)
+	rootShaVal := r.rootHash.Sum(nil)
+	data = encoding.EncodeBytes(data, []byte(fmt.Sprintf("%x", rootShaVal)))
+	writedLen, err = r.writer.Write(data)
+	if err != nil {
+		return fmt.Errorf("writing root node to repair tree file failure: %w", err)
+	}
+
+	// write footer
+	data = make([]byte, 0)
+	data = encoding.VarInt64ToBytes(data, leafNodesFinishedOffset) // end offset of the slot nodes
+	data = encoding.VarInt64ToBytes(data, int64(len(r.Slots)))     // count of the slot nodes
+	data = encoding.VarInt64ToBytes(data, slotNodesFinishedOffset) // end offset of the root node
+	data = encoding.VarInt64ToBytes(data, int64(writedLen))        // length of the root node
+	footerLen := uint8(len(data))
```

Comment:
No validation that the footer length fits within uint8 bounds, which could cause silent truncation.